### ISSUE TITLE
Replace label units for DBNs with a softmax unit

### ISSUE THREAD

**Issue by exilefaker** [exilefaker/python-UL#7]
Currently, I use an unprincipled procedure to constrain generated labels to be one-hot vectors: the label with maximum probability is turned on and the others are turned off. This representation is then used to drive CD learning.

It would be much better, as per https://www.cs.toronto.edu/~hinton/absps/guideTR.pdf, to use a softmax unit for the labels, and then use the log probability of the correct class as the gradient (i.e. cross-entropy loss).
